packages/components/src/slot-fill/context.js

```diff
@@ -21,22 +22,49 @@ class SlotFillProvider extends Component {
 	constructor() {
 		super( ...arguments );
 
-		this.registerSlot = this.registerSlot.bind( this );
-		this.registerFill = this.registerFill.bind( this );
-		this.unregisterSlot = this.unregisterSlot.bind( this );
-		this.unregisterFill = this.unregisterFill.bind( this );
-		this.getSlot = this.getSlot.bind( this );
-		this.getFills = this.getFills.bind( this );
-
 		this.slots = {};
 		this.fills = {};
 		this.state = {
-			registerSlot: this.registerSlot,
-			unregisterSlot: this.unregisterSlot,
-			registerFill: this.registerFill,
-			unregisterFill: this.unregisterFill,
-			getSlot: this.getSlot,
-			getFills: this.getFills,
+			registerSlot: this.proxy( 'registerSlot' ).bind( this ),
+			registerFill: this.proxy( 'registerFill' ).bind( this ),
+			unregisterSlot: this.proxy( 'unregisterSlot' ).bind( this ),
+			unregisterFill: this.proxy( 'unregisterFill' ).bind( this ),
+			getSlot: this.proxy( 'getSlot' ).bind( this ),
+			getFills: this.proxy( 'getFills' ).bind( this ),
+		};
+	}
+
+	/**
+	 * Given a function name for a function on the SlotFillProvider prototype,
+	 * returns a new function which either passes the arguments to the current
+	 * instance, or to the context ancestor, dependent on whether the provider
+	 * is configured to handle the given slot.
+	 *
+	 * @param {string} functionName SlotFillProvider function name.
+	 *
+	 * @return {Function} Proxying function.
+	 */
+	proxy( functionName ) {
+		return ( name, ...args ) => {
+			const { slots: handledSlots } = this.props;
+
+			let handler = this[ functionName ];
+
+			if ( Array.isArray( handledSlots ) ) {
+				const isHandled = some( handledSlots, ( slotNameOrComponent ) => {
+					const { slotName = slotNameOrComponent } = slotNameOrComponent;
+					return typeof slotName === 'string' && (
+						slotName === name ||
+						name.startsWith( slotName + '.' )
+					);
+				} );
+
+				if ( ! isHandled ) {
+					handler = this.props[ functionName ];
+				}
+			}
+
+			return handler.call( this, name, ...args );
```

Comment:
I'm probably missing something but is .call really required? Couldn't we just execute `return handler( name, ...args );`. Why do we have a need to change the "owner object"?

Reply:
> I'm probably missing something but is .call really required? Couldn't we just execute `return handler( name, ...args );`. Why do we have a need to change the "owner object"?

I don't immediately recall why or if this was necessary. Looking at it now, it feels like it may have been at least necessary that a `this` context exist in the component instance function being called, but as I see it here, it's probably wrong to consider it the `this` of the _current_ instance (rather than the ancestor's), and I would think even without `.call`, the `this` would still be assigned in the ancestor (since most components define instance properties with `.bind` in the constructor).